FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,46 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		context.perform { [context] in
+			do {
+				let request = NSFetchRequest<ManagedCache>(entityName: "\(ManagedCache.self)")
+				request.returnsObjectsAsFaults = false
+				if let cache = try ManagedCache.find(in: context) {
+					completion(.found(feed: cache.localFeed, timestamp: cache.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		context.perform { [context] in
+			do {
+				let cache = try ManagedCache.newUniqueInstance(in: context)
+				cache.timestamp = timestamp
+				cache.feed = ManagedFeedImage.images(from: feed, in: context)
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		context.perform { [context] in
+			do {
+				try ManagedCache.delete(context: context)
+				try context.save()
```

Comment:
If there's no cache stored in Core Data, `ManagedCache.delete` won't do anything. But `context.save()` should only be called when there are changes. If there are no changes, Core Data may perform unnecessary work.

One way you can optimize this part is to check if there are changes before calling `save`:

 ```suggestion
				try ManagedCache.delete(context: context)
				if context.hasChanges {
					try context.save()
				}
```